In [35]:
import warnings
warnings.filterwarnings("ignore")

from dask.distributed import Client, progress

client = Client(n_workers=4, threads_per_worker=2, memory_limit='2GB')
client

<Client: 'tcp://127.0.0.1:58012' processes=4 threads=8, memory=7.45 GiB>

In [36]:
# DataFrames implement the pandas API
import dask.dataframe as dd

# This loads the data into a Dask DataFrame
df = dd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/creditcard.csv')

In [37]:
print(df)

Dask DataFrame Structure:
                Time       V1       V2       V3       V4       V5       V6       V7       V8       V9      V10      V11      V12      V13      V14      V15      V16      V17      V18      V19      V20      V21      V22      V23      V24      V25      V26      V27      V28   Amount  Class
npartitions=3                                                                                                                                                                                                                                                                                   
               int64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  int64
                 ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      

In [38]:
df.dtypes

Time        int64
V1        float64
V2        float64
V3        float64
V4        float64
V5        float64
V6        float64
V7        float64
V8        float64
V9        float64
V10       float64
V11       float64
V12       float64
V13       float64
V14       float64
V15       float64
V16       float64
V17       float64
V18       float64
V19       float64
V20       float64
V21       float64
V22       float64
V23       float64
V24       float64
V25       float64
V26       float64
V27       float64
V28       float64
Amount    float64
Class       int64
dtype: object

In [39]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [40]:
df.describe()

# It looks like calling 'df.describe' and 'df.dtypes' we get different values for df['time']. If this is the case in the future, remember to add <dtype={'column_name': 'data_type'}> to the dd.read_csv parameters.

Dask DataFrame Structure:
                  Time       V1       V2       V3       V4       V5       V6       V7       V8       V9      V10      V11      V12      V13      V14      V15      V16      V17      V18      V19      V20      V21      V22      V23      V24      V25      V26      V27      V28   Amount    Class
npartitions=1                                                                                                                                                                                                                                                                                       
               float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...
Dask Name: describe-numeric, 344 tasks

In [41]:
df.describe().compute()

ValueError: Mismatched dtypes found in `pd.read_csv`/`pd.read_table`.

+--------+---------+----------+
| Column | Found   | Expected |
+--------+---------+----------+
| Time   | float64 | int64    |
+--------+---------+----------+

Usually this is due to dask's dtype inference failing, and
*may* be fixed by specifying dtypes manually by adding:

dtype={'Time': 'float64'}

to the call to `read_csv`/`read_table`.

Alternatively, provide `assume_missing=True` to interpret
all unspecified integer columns as floats.

In [42]:
df = dd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/creditcard.csv', dtype={'Time': 'float64'})


In [43]:
df.describe().compute()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  6.003539e-16  6.386744e-18 -3.078411e-15  2.484443e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     49346.000000 -7.868600e-01 -5.536365e-01 -7.156553e-01 -7.067841e-01   
50%     76029.000000  9.301532e-02  8.999754e-02  1.743346e-01  1.809941e-01   
75%    138472.000000  1.912886e+00  8.916958e-01  1.373682e+00  1.012508e+00   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   3.832046e-16  1.542399e-15 -1.341216e-15  1.532819e-16 -1.362372e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -4.093779e-01 -6.546632e-01 -4.787135e-01 -1.339191e-01 -5.477089e-01   
50%    1.480477e-01 -1.649056e-01  1.590763e-01  7.790153e-02  2.583745e-02   
75%    8.613822e-01  4.775488e-01  7.334868e-01  3.706835e-01  7.092123e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  3.512709e-17  1.119277e-15  3.153455e-16  4.394479e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.256367e-01 -5.250530e-01 -1.301482e-01 -3.241623e-01   
50%    ...  6.636310e-03  1.241855e-01  5.160336e-02  6.794774e-02   
75%    ...  2.385322e-01  7.308589e-01  2.346538e-01  5.274699e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   1.079360e-15  1.743781e-15 -3.506722e-16 -1.003916e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -2.218416e-01 -2.821043e-01 -6.089022e-02 -2.762231e-02       6.840000   
50%    1.674311e-01 -1.290092e-03  1.088230e-02  2.345040e-02      24.990000   
75%    4.190202e-01  2.938470e-01  1.060637e-01  8.061929e-02      84.910000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [44]:
df2 = df[df["Amount"] > 10000]
df

Dask DataFrame Structure:
                  Time       V1       V2       V3       V4       V5       V6       V7       V8       V9      V10      V11      V12      V13      V14      V15      V16      V17      V18      V19      V20      V21      V22      V23      V24      V25      V26      V27      V28   Amount  Class
npartitions=3                                                                                                                                                                                                                                                                                     
               float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  int64
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
Dask Name: read-csv, 3 tasks

In [45]:
df[df["Amount"] > 15000].compute()

Time         V1         V2         V3         V4          V5  \
58465   48401.0 -36.802320 -63.344698 -20.645794  16.715537  -20.672064   
30325   95286.0 -34.549296 -60.464618 -21.340854  16.875344  -19.229075   
32983  166198.0 -35.548539 -31.850484 -48.325589  15.304184 -113.743307   

              V6          V7         V8        V9  ...        V21        V22  \
58465   7.694002   24.956587  -4.730111 -2.687312  ...  11.455313 -10.933144   
30325   6.335259   24.422716  -4.964566  0.188912  ...  11.502580  -9.499423   
32983  73.301626  120.589494 -27.347360 -3.872425  ... -21.620120   5.712303   

             V23       V24       V25       V26        V27        V28  \
58465 -17.173665  1.180700 -7.025783 -2.534330  -3.602479   3.450224   
30325 -16.513186  0.744341 -7.081325 -2.604551  -3.550963   3.250802   
32983  -1.581098  4.584549  4.554683  3.415636  31.612198 -15.430084   

         Amount  Class  
58465  19656.53      0  
30325  18910.00      0  
32983  25691.16      0  

[3 rows x 31 columns]

In [46]:
df.loc[df["Amount"] > 10000,:].compute()

Time         V1         V2         V3         V4          V5  \
46841   42951.0 -23.712839 -42.172688 -13.320825   9.925019  -13.945538   
54018   46253.0 -21.780665 -38.305310 -12.122469   9.752791  -12.880794   
58465   48401.0 -36.802320 -63.344698 -20.645794  16.715537  -20.672064   
30325   95286.0 -34.549296 -60.464618 -21.340854  16.875344  -19.229075   
48486  119713.0 -20.924897 -37.943452 -14.060281  10.473005  -10.866639   
32983  166198.0 -35.548539 -31.850484 -48.325589  15.304184 -113.743307   
42461  172273.0  -9.030538 -11.112584 -16.233798   3.592021  -40.427726   

              V6          V7         V8        V9  ...        V21        V22  \
46841   5.564891   15.710644  -2.844253 -1.580725  ...   7.921600  -6.320710   
54018   4.256017   14.785051  -2.818253 -0.667338  ...   7.437478  -5.619439   
58465   7.694002   24.956587  -4.730111 -2.687312  ...  11.455313 -10.933144   
30325   6.335259   24.422716  -4.964566  0.188912  ...  11.502580  -9.499423   
48486   6.256654   14.960521  -2.392155 -0.597076  ...   6.829810  -6.926353   
32983  73.301626  120.589494 -27.347360 -3.872425  ... -21.620120   5.712303   
42461  23.917837   44.054461  -7.277778 -4.210637  ...  -0.269048   0.988144   

             V23       V24       V25       V26        V27        V28  \
46841 -11.310338  0.404175 -4.547278 -1.577118  -2.357385   2.253662   
54018 -10.547038  0.653249 -4.232409 -0.480459  -2.257913   2.082488   
58465 -17.173665  1.180700 -7.025783 -2.534330  -3.602479   3.450224   
30325 -16.513186  0.744341 -7.081325 -2.604551  -3.550963   3.250802   
48486  -9.928657 -0.447084 -4.848151 -2.241620  -2.140723   2.001492   
32983  -1.581098  4.584549  4.554683  3.415636  31.612198 -15.430084   
42461   7.040028  0.347693  2.520869  2.342495   3.478175  -2.713136   

         Amount  Class  
46841  12910.93      0  
54018  11898.09      0  
58465  19656.53      0  
30325  18910.00      0  
48486  11789.84      0  
32983  25691.16      0  
42461  10199.44      0  

[7 rows x 31 columns]

In [47]:
# Trying to count all rows in the Dask DataFrame
df.shape[0].compute()

284807

In [48]:
df2 = df.groupby("Class")["Amount"].mean().compute()
df2

Class
0     88.291022
1    122.211321
Name: Amount, dtype: float64

In [49]:
df["Amount2"] = df["Amount"].apply(lambda x: 1 if x>10000 else 0, meta=('Amount2', 'int64'))
df["Amount2"].head(20)

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
16    0
17    0
18    0
19    0
Name: Amount2, dtype: int64

In [50]:
df = df.persist()

# Assignments
## 1. How many transactions are there in total:

In [65]:
# This is another method other than looking as [0] of .shape called on the df. This confirms how many non-null rows there are in each column
df.count().compute()

Time       284807
V1         284807
V2         284807
V3         284807
V4         284807
V5         284807
V6         284807
V7         284807
V8         284807
V9         284807
V10        284807
V11        284807
V12        284807
V13        284807
V14        284807
V15        284807
V16        284807
V17        284807
V18        284807
V19        284807
V20        284807
V21        284807
V22        284807
V23        284807
V24        284807
V25        284807
V26        284807
V27        284807
V28        284807
Amount     284807
Class      284807
Amount2    284807
dtype: int64

In [64]:
# The below would do the trick too
len(df)

284807

## Answer: 284807

## 2. How many transactions are fraud, and how many are not fraud?:

In [54]:
# Counting the number of rows for each class.
df.groupby("Class").count().compute()

Time      V1      V2      V3      V4      V5      V6      V7      V8  \
Class                                                                           
0      284315  284315  284315  284315  284315  284315  284315  284315  284315   
1         492     492     492     492     492     492     492     492     492   

           V9  ...     V21     V22     V23     V24     V25     V26     V27  \
Class          ...                                                           
0      284315  ...  284315  284315  284315  284315  284315  284315  284315   
1         492  ...     492     492     492     492     492     492     492   

          V28  Amount  Amount2  
Class                           
0      284315  284315   284315  
1         492     492      492  

[2 rows x 31 columns]

## Answer: Not fraud - 284315, Fraud - 492

## 3. What is the maximum amount in fraud transactions?:

In [61]:
# Selecting the .max() value from the "Amount" column of the df where the "Class" == 1.
df[df.Class == 1]["Amount"].max().compute()

2125.87

## Answer: $2125.87